File: LeetCode/766.Toeplitz_Matrix/Mario/766.Toeplitz_Matrix.py
#!/usr/bin/env python3


class Solution:
    def isToeplitzMatrix(self, matrix: List[List[int]]) -> bool:
        # i: [0, m-1] j:[]
        # i - j = fixed num [m-1, -(n-1)]

        m, n = len(matrix), len(matrix[0])

        dic = defaultdict(int)
        for i in range(m):
            for j in range(n):
                key = i - j
                if key not in dic:
                    dic[key] = matrix[i][j]
                elif dic[key] != matrix[i][j]:
                    return False

        return True

File: LeetCode/766.Toeplitz_Matrix/Mario/766.Toeplitz_Matrix.swift
class Solution {
    func isToeplitzMatrix(_ matrix: [[Int]]) -> Bool {
        let m = matrix.count, n = matrix[0].count

        var dic: [Int: Int] = [:]

        for i in 0 ..< m {
            for j in 0 ..< n {
                let key = i - j
                if let val = dic[key] {
                    if val != matrix[i][j] {
                        return false
                    }
                } else {
                    dic[key] = matrix[i][j]
                }
            }
        }

        return true
    }
}

class Solution {
    func isToeplitzMatrix(_ matrix: [[Int]]) -> Bool {
        let m = matrix.count, n = matrix[0].count

        for i in 1 ..< m {
            for j in 1 ..< n {
                if matrix[i][j] != matrix[i - 1][j - 1] {
                    return false
                }
            }
        }

        return true
    }
}
